File: packages/go/hushspec/receipt.go
package hushspec

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"strings"
	"time"
)

// DecisionReceipt is an auditable record of a single policy evaluation.
type DecisionReceipt struct {
	ReceiptID            string            `json:"receipt_id"`
	Timestamp            string            `json:"timestamp"`
	HushSpecVersion      string            `json:"hushspec_version"`
	Action               ActionSummary     `json:"action"`
	Decision             Decision          `json:"decision"`
	MatchedRule          string            `json:"matched_rule,omitempty"`
	Reason               string            `json:"reason,omitempty"`
	RuleTrace            []RuleEvaluation  `json:"rule_trace"`
	Policy               PolicySummary     `json:"policy"`
	OriginProfile        string            `json:"origin_profile,omitempty"`
	Posture              *PostureResult    `json:"posture,omitempty"`
	EvaluationDurationUs int64             `json:"evaluation_duration_us"`
}

type ActionSummary struct {
	Type            string `json:"type"`
	Target          string `json:"target,omitempty"`
	ContentRedacted bool   `json:"content_redacted,omitempty"`
}

type RuleOutcome string

const (
	RuleOutcomeAllow RuleOutcome = "allow"
	RuleOutcomeWarn  RuleOutcome = "warn"
	RuleOutcomeDeny  RuleOutcome = "deny"
	RuleOutcomeSkip  RuleOutcome = "skip"
)

type RuleEvaluation struct {
	RuleBlock   string      `json:"rule_block"`
	Outcome     RuleOutcome `json:"outcome"`
	MatchedRule string      `json:"matched_rule,omitempty"`
	Reason      string      `json:"reason,omitempty"`
	Evaluated   bool        `json:"evaluated"`
}

type PolicySummary struct {
	Name        string `json:"name,omitempty"`
	Version     string `json:"version"`
	ContentHash string `json:"content_hash"`
}

// AuditConfig controls receipt verbosity. When Enabled is false, the receipt
// contains the correct decision but skips timing, rule trace, and policy hashing.
type AuditConfig struct {
	Enabled          bool
	IncludeRuleTrace bool
	RedactContent    bool
}

// DefaultAuditConfig returns an AuditConfig with all features enabled.
func DefaultAuditConfig() AuditConfig {
	return AuditConfig{
		Enabled:          true,
		IncludeRuleTrace: true,
		RedactContent:    true,
	}
}

// EvaluateAudited wraps Evaluate with timing, rule trace collection, and
// policy hashing, returning a full DecisionReceipt.
func EvaluateAudited(spec *HushSpec, action *EvaluationAction, config *AuditConfig) DecisionReceipt {
	var start time.Time
	if config.Enabled {
		start = time.Now()
	}

	result := Evaluate(spec, action)

	var durationUs int64
	if config.Enabled {
		durationUs = time.Since(start).Microseconds()
	}

	var ruleTrace []RuleEvaluation
	if config.Enabled && config.IncludeRuleTrace {
		ruleTrace = collectRuleTrace(spec, action, &result)
	} else {
		ruleTrace = []RuleEvaluation{}
	}

	var policy PolicySummary
	if config.Enabled {
		policy = buildPolicySummary(spec)
	} else {
		policy = PolicySummary{
			Name:        spec.Name,
			Version:     spec.HushSpecVersion,
			ContentHash: "",
		}
	}

	actionSummary := ActionSummary{
		Type:            action.Type,
		Target:          action.Target,
		ContentRedacted: config.RedactContent && action.Content != "",
	}

	return DecisionReceipt{
		ReceiptID:            generateUUIDv4(),
		Timestamp:            time.Now().UTC().Format(time.RFC3339Nano),
		HushSpecVersion:      Version,
		Action:               actionSummary,
		Decision:             result.Decision,
		MatchedRule:          result.MatchedRule,
		Reason:               result.Reason,
		RuleTrace:            ruleTrace,
		Policy:               policy,
		OriginProfile:        result.OriginProfile,
		Posture:              result.Posture,
		EvaluationDurationUs: durationUs,
	}
}

// ComputePolicyHash returns the SHA-256 hex digest of the JSON-serialized spec.
func ComputePolicyHash(spec *HushSpec) string {
	jsonBytes, err := json.Marshal(spec)
	if err != nil {
		return ""
	}
	hash := sha256.Sum256(jsonBytes)
	return fmt.Sprintf("%x", hash[:])
}

func generateUUIDv4() string {
	var uuid [16]byte
	_, _ = rand.Read(uuid[:])
	uuid[6] = (uuid[6] & 0x0f) | 0x40 // version 4
	uuid[8] = (uuid[8] & 0x3f) | 0x80 // variant 10
	return fmt.Sprintf("%x-%x-%x-%x-%x",
		uuid[0:4], uuid[4:6], uuid[6:8], uuid[8:10], uuid[10:16])
}

func buildPolicySummary(spec *HushSpec) PolicySummary {
	return PolicySummary{
		Name:        spec.Name,
		Version:     spec.HushSpecVersion,
		ContentHash: ComputePolicyHash(spec),
	}
}

func outcomeFromDecision(decision Decision) RuleOutcome {
	switch decision {
	case DecisionAllow:
		return RuleOutcomeAllow
	case DecisionWarn:
		return RuleOutcomeWarn
	case DecisionDeny:
		return RuleOutcomeDeny
	default:
		return RuleOutcomeAllow
	}
}

func collectRuleTrace(
	spec *HushSpec,
	action *EvaluationAction,
	result *EvaluationResult,
) []RuleEvaluation {
	var trace []RuleEvaluation

	if result.Posture != nil {
		postureDenied := result.MatchedRule != "" &&
			strings.HasPrefix(result.MatchedRule, "extensions.posture.states.")

		if postureDenied {
			trace = append(trace, RuleEvaluation{
				RuleBlock:   "posture_capability",
				Outcome:     RuleOutcomeDeny,
				MatchedRule: result.MatchedRule,
				Reason:      result.Reason,
				Evaluated:   true,
			})
			appendSkippedRules(action, &trace, "short-circuited by posture deny")
			return trace
		}

		trace = append(trace, RuleEvaluation{
			RuleBlock: "posture_capability",
			Outcome:   RuleOutcomeAllow,
			Reason:    "posture capabilities satisfied",
			Evaluated: true,
		})
	}

	switch action.Type {
	case "tool_call":
		traceToolAccess(spec, action, result, &trace)
	case "egress":
		traceEgressBlock(spec, result, &trace)
	case "file_read":
		tracePathGuards(spec, result, &trace)
	case "file_write":
		tracePathGuards(spec, result, &trace)
		traceSecretPatterns(spec, result, &trace)
	case "patch_apply":
		tracePathGuards(spec, result, &trace)
		tracePatchIntegrity(spec, result, &trace)
	case "shell_command":
		traceShellCommands(spec, result, &trace)
	case "computer_use":
		traceComputerUse(spec, result, &trace)
	default:
		trace = append(trace, RuleEvaluation{
			RuleBlock:   "default",
			Outcome:     outcomeFromDecision(result.Decision),
			MatchedRule: result.MatchedRule,
			Reason:      result.Reason,
			Evaluated:   true,
		})
	}

	return trace
}

func traceToolAccess(
	spec *HushSpec,
	action *EvaluationAction,
	result *EvaluationResult,
	trace *[]RuleEvaluation,
) {
	hasOriginRule := result.MatchedRule != "" &&
		strings.HasPrefix(result.MatchedRule, "extensions.origins.profiles.")

	hasRule := hasOriginRule ||
		(spec.Rules != nil && spec.Rules.ToolAccess != nil) ||
		action.Origin != nil

	if hasRule {
		*trace = append(*trace, RuleEvaluation{
			RuleBlock:   "tool_access",
			Outcome:     outcomeFromDecision(result.Decision),
			MatchedRule: result.MatchedRule,
			Reason:      result.Reason,
			Evaluated:   true,
		})
	} else {
		*trace = append(*trace, RuleEvaluation{
			RuleBlock: "tool_access",
			Outcome:   RuleOutcomeSkip,
			Reason:    "no tool_access rule configured",
			Evaluated: false,
		})
	}
}

func traceEgressBlock(
	spec *HushSpec,
	result *EvaluationResult,
	trace *[]RuleEvaluation,
) {
	hasOriginRule := result.MatchedRule != "" &&
		strings.Contains(result.MatchedRule, "egress")

	hasRule := hasOriginRule ||
		(spec.Rules != nil && spec.Rules.Egress != nil)

	if hasRule {
		*trace = append(*trace, RuleEvaluation{
			RuleBlock:   "egress",
			Outcome:     outcomeFromDecision(result.Decision),
			MatchedRule: result.MatchedRule,
			Reason:      result.Reason,
			Evaluated:   true,
		})
	} else {
		*trace = append(*trace, RuleEvaluation{
			RuleBlock: "egress",
			Outcome:   RuleOutcomeSkip,
			Reason:    "no egress rule configured",
			Evaluated: false,
		})
	}
}

func tracePathGuards(
	spec *HushSpec,
	result *EvaluationResult,
	trace *[]RuleEvaluation,
) {
	rules := spec.Rules
	decidedByForbidden := result.MatchedRule != "" &&
		strings.Contains(result.MatchedRule, "forbidden_paths")
	decidedByAllowlist := result.MatchedRule != "" &&
		strings.Contains(result.MatchedRule, "path_allowlist")

	if rules != nil && rules.ForbiddenPaths != nil {
		fp := rules.ForbiddenPaths
		if fp.Enabled {
			if decidedByForbidden {
				*trace = append(*trace, RuleEvaluation{
					RuleBlock:   "forbidden_paths",
					Outcome:     outcomeFromDecision(result.Decision),
					MatchedRule: result.MatchedRule,
					Reason:      result.Reason,
					Evaluated:   true,
				})
			} else {
				*trace = append(*trace, RuleEvaluation{
					RuleBlock: "forbidden_paths",
					Outcome:   RuleOutcomeAllow,
					Reason:    "path did not match any forbidden pattern",
					Evaluated: true,
				})
			}
		} else {
			*trace = append(*trace, RuleEvaluation{
				RuleBlock: "forbidden_paths",
				Outcome:   RuleOutcomeSkip,
				Reason:    "rule disabled",
				Evaluated: false,
			})
		}
	}

	if rules != nil && rules.PathAllowlist != nil {
		pa := rules.PathAllowlist
		if pa.Enabled {
			shortCircuited := decidedByForbidden && result.Decision == DecisionDeny
			if shortCircuited {
				*trace = append(*trace, RuleEvaluation{
					RuleBlock: "path_allowlist",
					Outcome:   RuleOutcomeSkip,
					Reason:    "short-circuited by prior deny",
					Evaluated: false,
				})
			} else if decidedByAllowlist {
				*trace = append(*trace, RuleEvaluation{
					RuleBlock:   "path_allowlist",
					Outcome:     outcomeFromDecision(result.Decision),
					MatchedRule: result.MatchedRule,
					Reason:      result.Reason,
					Evaluated:   true,
				})
			} else {
				*trace = append(*trace, RuleEvaluation{
					RuleBlock: "path_allowlist",
					Outcome:   RuleOutcomeAllow,
					Reason:    "path matched allowlist",
					Evaluated: true,
				})
			}
		} else {
			*trace = append(*trace, RuleEvaluation{
				RuleBlock: "path_allowlist",
				Outcome:   RuleOutcomeSkip,
				Reason:    "rule disabled",
				Evaluated: false,
			})
		}
	}
}

func traceSecretPatterns(
	spec *HushSpec,
	result *EvaluationResult,
	trace *[]RuleEvaluation,
) {
	decidedBySecret := result.MatchedRule != "" &&
		strings.Contains(result.MatchedRule, "secret_patterns")
	priorDeny := hasPriorDeny(*trace)

	if spec.Rules != nil && spec.Rules.SecretPatterns != nil {
		sp := spec.Rules.SecretPatterns
		if sp.Enabled {
			if priorDeny {
				*trace = append(*trace, RuleEvaluation{
					RuleBlock: "secret_patterns",
					Outcome:   RuleOutcomeSkip,
					Reason:    "short-circuited by prior deny",
					Evaluated: false,
				})
			} else if decidedBySecret {
				*trace = append(*trace, RuleEvaluation{
					RuleBlock:   "secret_patterns",
					Outcome:     outcomeFromDecision(result.Decision),
					MatchedRule: result.MatchedRule,
					Reason:      result.Reason,
					Evaluated:   true,
				})
			} else {
				*trace = append(*trace, RuleEvaluation{
					RuleBlock: "secret_patterns",
					Outcome:   RuleOutcomeAllow,
					Reason:    "content did not match any secret pattern",
					Evaluated: true,
				})
			}
		} else {
			*trace = append(*trace, RuleEvaluation{
				RuleBlock: "secret_patterns",
				Outcome:   RuleOutcomeSkip,
				Reason:    "rule disabled",
				Evaluated: false,
			})
		}
	}
}

func tracePatchIntegrity(
	spec *HushSpec,
	result *EvaluationResult,
	trace *[]RuleEvaluation,
) {
	decidedByPatch := result.MatchedRule != "" &&
		strings.Contains(result.MatchedRule, "patch_integrity")
	priorDeny := hasPriorDeny(*trace)

	if spec.Rules != nil && spec.Rules.PatchIntegrity != nil {
		pi := spec.Rules.PatchIntegrity
		if pi.Enabled {
			if priorDeny {
				*trace = append(*trace, RuleEvaluation{
					RuleBlock: "patch_integrity",
					Outcome:   RuleOutcomeSkip,
					Reason:    "short-circuited by prior deny",
					Evaluated: false,
				})
			} else if decidedByPatch {
				*trace = append(*trace, RuleEvaluation{
					RuleBlock:   "patch_integrity",
					Outcome:     outcomeFromDecision(result.Decision),
					MatchedRule: result.MatchedRule,
					Reason:      result.Reason,
					Evaluated:   true,
				})
			} else {
				*trace = append(*trace, RuleEvaluation{
					RuleBlock: "patch_integrity",
					Outcome:   RuleOutcomeAllow,
					Reason:    "patch passed integrity checks",
					Evaluated: true,
				})
			}
		} else {
			*trace = append(*trace, RuleEvaluation{
				RuleBlock: "patch_integrity",
				Outcome:   RuleOutcomeSkip,
				Reason:    "rule disabled",
				Evaluated: false,
			})
		}
	}
}

func traceShellCommands(
	spec *HushSpec,
	result *EvaluationResult,
	trace *[]RuleEvaluation,
) {
	decidedByShell := result.MatchedRule != "" &&
		strings.Contains(result.MatchedRule, "shell_commands")

	if spec.Rules != nil && spec.Rules.ShellCommands != nil {
		sc := spec.Rules.ShellCommands
		if sc.Enabled {
			if decidedByShell {
				*trace = append(*trace, RuleEvaluation{
					RuleBlock:   "shell_commands",
					Outcome:     outcomeFromDecision(result.Decision),
					MatchedRule: result.MatchedRule,
					Reason:      result.Reason,
					Evaluated:   true,
				})
			} else {
				*trace = append(*trace, RuleEvaluation{
					RuleBlock: "shell_commands",
					Outcome:   RuleOutcomeAllow,
					Reason:    "command did not match any forbidden pattern",
					Evaluated: true,
				})
			}
		} else {
			*trace = append(*trace, RuleEvaluation{
				RuleBlock: "shell_commands",
				Outcome:   RuleOutcomeSkip,
				Reason:    "rule disabled",
				Evaluated: false,
			})
		}
	}
}

func traceComputerUse(
	spec *HushSpec,
	result *EvaluationResult,
	trace *[]RuleEvaluation,
) {
	decidedByComputer := result.MatchedRule != "" &&
		strings.Contains(result.MatchedRule, "computer_use")

	if spec.Rules != nil && spec.Rules.ComputerUse != nil {
		cu := spec.Rules.ComputerUse
		if cu.Enabled {
			if decidedByComputer {
				*trace = append(*trace, RuleEvaluation{
					RuleBlock:   "computer_use",
					Outcome:     outcomeFromDecision(result.Decision),
					MatchedRule: result.MatchedRule,
					Reason:      result.Reason,
					Evaluated:   true,
				})
			} else {
				*trace = append(*trace, RuleEvaluation{
					RuleBlock: "computer_use",
					Outcome:   RuleOutcomeAllow,
					Reason:    "action allowed by computer_use rule",
					Evaluated: true,
				})
			}
		} else {
			*trace = append(*trace, RuleEvaluation{
				RuleBlock: "computer_use",
				Outcome:   RuleOutcomeSkip,
				Reason:    "rule disabled",
				Evaluated: false,
			})
		}
	}
}

func hasPriorDeny(trace []RuleEvaluation) bool {
	for _, t := range trace {
		if t.Outcome == RuleOutcomeDeny && t.Evaluated {
			return true
		}
	}
	return false
}

func appendSkippedRules(
	action *EvaluationAction,
	trace *[]RuleEvaluation,
	reason string,
) {
	var blocks []string
	switch action.Type {
	case "tool_call":
		blocks = []string{"tool_access"}
	case "egress":
		blocks = []string{"egress"}
	case "file_read":
		blocks = []string{"forbidden_paths", "path_allowlist"}
	case "file_write":
		blocks = []string{"forbidden_paths", "path_allowlist", "secret_patterns"}
	case "patch_apply":
		blocks = []string{"forbidden_paths", "path_allowlist", "patch_integrity"}
	case "shell_command":
		blocks = []string{"shell_commands"}
	case "computer_use":
		blocks = []string{"computer_use"}
	}

	for _, block := range blocks {
		*trace = append(*trace, RuleEvaluation{
			RuleBlock: block,
			Outcome:   RuleOutcomeSkip,
			Reason:    reason,
			Evaluated: false,
		})
	}
}

File: packages/hushspec/src/receipt.ts
import { createHash, randomUUID } from 'node:crypto';
import type { HushSpec } from './schema.js';
import type { PostureResult } from './evaluate.js';
import { evaluate } from './evaluate.js';
import { HUSHSPEC_VERSION } from './version.js';
import type { EvaluationAction, EvaluationResult, Decision } from './evaluate.js';

export interface DecisionReceipt {
  receipt_id: string;
  timestamp: string;
  hushspec_version: string;
  action: ActionSummary;
  decision: Decision;
  matched_rule?: string;
  reason?: string;
  rule_trace: RuleEvaluation[];
  policy: PolicySummary;
  origin_profile?: string;
  posture?: PostureResult;
  evaluation_duration_us: number;
}

export interface ActionSummary {
  type: string;
  target?: string;
  content_redacted: boolean;
}

export type RuleOutcome = 'allow' | 'warn' | 'deny' | 'skip';

export interface RuleEvaluation {
  rule_block: string;
  outcome: RuleOutcome;
  matched_rule?: string;
  reason?: string;
  evaluated: boolean;
}

export interface PolicySummary {
  name?: string;
  version: string;
  /** SHA-256 hex digest of the canonical JSON serialization. */
  content_hash: string;
}

export interface AuditConfig {
  /** When false, skip timing, rule tracing, and policy hashing. */
  enabled: boolean;
  include_rule_trace: boolean;
  redact_content: boolean;
}

export const DEFAULT_AUDIT_CONFIG: AuditConfig = {
  enabled: true,
  include_rule_trace: true,
  redact_content: true,
};

export function evaluateAudited(
  spec: HushSpec,
  action: EvaluationAction,
  config: AuditConfig,
): DecisionReceipt {
  const startHr = config.enabled ? performance.now() : 0;
  const result = evaluate(spec, action);

  const durationUs = config.enabled
    ? Math.round((performance.now() - startHr) * 1000)
    : 0;

  const ruleTrace =
    config.enabled && config.include_rule_trace
      ? collectRuleTrace(spec, action, result)
      : [];

  const policy: PolicySummary = config.enabled
    ? buildPolicySummary(spec)
    : {
        name: spec.name,
        version: spec.hushspec,
        content_hash: '',
      };

  const actionSummary: ActionSummary = {
    type: action.type,
    target: action.target,
    content_redacted: config.redact_content && action.content != null,
  };

  return {
    receipt_id: randomUUID(),
    timestamp: new Date().toISOString(),
    hushspec_version: HUSHSPEC_VERSION,
    action: actionSummary,
    decision: result.decision,
    matched_rule: result.matched_rule,
    reason: result.reason,
    rule_trace: ruleTrace,
    policy,
    origin_profile: result.origin_profile,
    posture: result.posture,
    evaluation_duration_us: durationUs,
  };
}

export function computePolicyHash(spec: HushSpec): string {
  const json = JSON.stringify(spec);
  return createHash('sha256').update(json).digest('hex');
}

function buildPolicySummary(spec: HushSpec): PolicySummary {
  return {
    name: spec.name,
    version: spec.hushspec,
    content_hash: computePolicyHash(spec),
  };
}

function collectRuleTrace(
  spec: HushSpec,
  action: EvaluationAction,
  result: EvaluationResult,
): RuleEvaluation[] {
  const trace: RuleEvaluation[] = [];

  if (result.posture != null) {
    const postureDenied =
      result.matched_rule != null &&
      result.matched_rule.startsWith('extensions.posture.states.');

    if (postureDenied) {
      trace.push({
        rule_block: 'posture_capability',
        outcome: 'deny',
        matched_rule: result.matched_rule,
        reason: result.reason,
        evaluated: true,
      });
      appendSkippedRules(action, trace, 'short-circuited by posture deny');
      return trace;
    }

    trace.push({
      rule_block: 'posture_capability',
      outcome: 'allow',
      matched_rule: undefined,
      reason: 'posture capabilities satisfied',
      evaluated: true,
    });
  }

  switch (action.type) {
    case 'tool_call':
      traceToolAccess(spec, action, result, trace);
      break;
    case 'egress':
      traceEgress(spec, result, trace);
      break;
    case 'file_read':
      tracePathGuards(spec, result, trace);
      break;
    case 'file_write':
      tracePathGuards(spec, result, trace);
      traceSecretPatterns(spec, result, trace);
      break;
    case 'patch_apply':
      tracePathGuards(spec, result, trace);
      tracePatchIntegrity(spec, result, trace);
      break;
    case 'shell_command':
      traceShellCommands(spec, result, trace);
      break;
    case 'computer_use':
      traceComputerUse(spec, result, trace);
      break;
    default:
      trace.push({
        rule_block: 'default',
        outcome: result.decision,
        matched_rule: result.matched_rule,
        reason: result.reason,
        evaluated: true,
      });
      break;
  }

  return trace;
}

function traceToolAccess(
  spec: HushSpec,
  action: EvaluationAction,
  result: EvaluationResult,
  trace: RuleEvaluation[],
): void {
  const hasOriginRule =
    result.matched_rule != null &&
    result.matched_rule.startsWith('extensions.origins.profiles.');

  const hasRule =
    hasOriginRule ||
    spec.rules?.tool_access != null ||
    action.origin != null;

  if (hasRule) {
    trace.push({
      rule_block: 'tool_access',
      outcome: result.decision,
      matched_rule: result.matched_rule,
      reason: result.reason,
      evaluated: true,
    });
  } else {
    trace.push({
      rule_block: 'tool_access',
      outcome: 'skip',
      matched_rule: undefined,
      reason: 'no tool_access rule configured',
      evaluated: false,
    });
  }
}

function traceEgress(
  spec: HushSpec,
  result: EvaluationResult,
  trace: RuleEvaluation[],
): void {
  const hasOriginRule =
    result.matched_rule != null &&
    result.matched_rule.includes('egress');

  const hasRule = hasOriginRule || spec.rules?.egress != null;

  if (hasRule) {
    trace.push({
      rule_block: 'egress',
      outcome: result.decision,
      matched_rule: result.matched_rule,
      reason: result.reason,
      evaluated: true,
    });
  } else {
    trace.push({
      rule_block: 'egress',
      outcome: 'skip',
      matched_rule: undefined,
      reason: 'no egress rule configured',
      evaluated: false,
    });
  }
}

function tracePathGuards(
  spec: HushSpec,
  result: EvaluationResult,
  trace: RuleEvaluation[],
): void {
  const rules = spec.rules;
  const decidedByForbidden =
    result.matched_rule != null &&
    result.matched_rule.includes('forbidden_paths');
  const decidedByAllowlist =
    result.matched_rule != null &&
    result.matched_rule.includes('path_allowlist');

  const fp = rules?.forbidden_paths;
  if (fp != null) {
    if (fp.enabled !== false) {
      if (decidedByForbidden) {
        trace.push({
          rule_block: 'forbidden_paths',
          outcome: result.decision,
          matched_rule: result.matched_rule,
          reason: result.reason,
          evaluated: true,
        });
      } else {
        trace.push({
          rule_block: 'forbidden_paths',
          outcome: 'allow',
          matched_rule: undefined,
          reason: 'path did not match any forbidden pattern',
          evaluated: true,
        });
      }
    } else {
      trace.push({
        rule_block: 'forbidden_paths',
        outcome: 'skip',
        matched_rule: undefined,
        reason: 'rule disabled',
        evaluated: false,
      });
    }
  }

  const pa = rules?.path_allowlist;
  if (pa != null) {
    if (pa.enabled !== false) {
      const shortCircuited =
        decidedByForbidden && result.decision === 'deny';
      if (shortCircuited) {
        trace.push({
          rule_block: 'path_allowlist',
          outcome: 'skip',
          matched_rule: undefined,
          reason: 'short-circuited by prior deny',
          evaluated: false,
        });
      } else if (decidedByAllowlist) {
        trace.push({
          rule_block: 'path_allowlist',
          outcome: result.decision,
          matched_rule: result.matched_rule,
          reason: result.reason,
          evaluated: true,
        });
      } else {
        trace.push({
          rule_block: 'path_allowlist',
          outcome: 'allow',
          matched_rule: undefined,
          reason: 'path matched allowlist',
          evaluated: true,
        });
      }
    } else {
      trace.push({
        rule_block: 'path_allowlist',
        outcome: 'skip',
        matched_rule: undefined,
        reason: 'rule disabled',
        evaluated: false,
      });
    }
  }
}

function traceSecretPatterns(
  spec: HushSpec,
  result: EvaluationResult,
  trace: RuleEvaluation[],
): void {
  const decidedBySecret =
    result.matched_rule != null &&
    result.matched_rule.includes('secret_patterns');
  const priorDeny = trace.some(
    (t) => t.outcome === 'deny' && t.evaluated,
  );

  const sp = spec.rules?.secret_patterns;
  if (sp != null) {
    if (sp.enabled !== false) {
      if (priorDeny) {
        trace.push({
          rule_block: 'secret_patterns',
          outcome: 'skip',
          matched_rule: undefined,
          reason: 'short-circuited by prior deny',
          evaluated: false,
        });
      } else if (decidedBySecret) {
        trace.push({
          rule_block: 'secret_patterns',
          outcome: result.decision,
          matched_rule: result.matched_rule,
          reason: result.reason,
          evaluated: true,
        });
      } else {
        trace.push({
          rule_block: 'secret_patterns',
          outcome: 'allow',
          matched_rule: undefined,
          reason: 'content did not match any secret pattern',
          evaluated: true,
        });
      }
    } else {
      trace.push({
        rule_block: 'secret_patterns',
        outcome: 'skip',
        matched_rule: undefined,
        reason: 'rule disabled',
        evaluated: false,
      });
    }
  }
}

function tracePatchIntegrity(
  spec: HushSpec,
  result: EvaluationResult,
  trace: RuleEvaluation[],
): void {
  const decidedByPatch =
    result.matched_rule != null &&
    result.matched_rule.includes('patch_integrity');
  const priorDeny = trace.some(
    (t) => t.outcome === 'deny' && t.evaluated,
  );

  const pi = spec.rules?.patch_integrity;
  if (pi != null) {
    if (pi.enabled !== false) {
      if (priorDeny) {
        trace.push({
          rule_block: 'patch_integrity',
          outcome: 'skip',
          matched_rule: undefined,
          reason: 'short-circuited by prior deny',
          evaluated: false,
        });
      } else if (decidedByPatch) {
        trace.push({
          rule_block: 'patch_integrity',
          outcome: result.decision,
          matched_rule: result.matched_rule,
          reason: result.reason,
          evaluated: true,
        });
      } else {
        trace.push({
          rule_block: 'patch_integrity',
          outcome: 'allow',
          matched_rule: undefined,
          reason: 'patch passed integrity checks',
          evaluated: true,
        });
      }
    } else {
      trace.push({
        rule_block: 'patch_integrity',
        outcome: 'skip',
        matched_rule: undefined,
        reason: 'rule disabled',
        evaluated: false,
      });
    }
  }
}

function traceShellCommands(
  spec: HushSpec,
  result: EvaluationResult,
  trace: RuleEvaluation[],
): void {
  const decidedByShell =
    result.matched_rule != null &&
    result.matched_rule.includes('shell_commands');

  const sc = spec.rules?.shell_commands;
  if (sc != null) {
    if (sc.enabled !== false) {
      if (decidedByShell) {
        trace.push({
          rule_block: 'shell_commands',
          outcome: result.decision,
          matched_rule: result.matched_rule,
          reason: result.reason,
          evaluated: true,
        });
      } else {
        trace.push({
          rule_block: 'shell_commands',
          outcome: 'allow',
          matched_rule: undefined,
          reason: 'command did not match any forbidden pattern',
          evaluated: true,
        });
      }
    } else {
      trace.push({
        rule_block: 'shell_commands',
        outcome: 'skip',
        matched_rule: undefined,
        reason: 'rule disabled',
        evaluated: false,
      });
    }
  }
}

function traceComputerUse(
  spec: HushSpec,
  result: EvaluationResult,
  trace: RuleEvaluation[],
): void {
  const decidedByComputer =
    result.matched_rule != null &&
    result.matched_rule.includes('computer_use');

  const cu = spec.rules?.computer_use;
  if (cu != null) {
    if (cu.enabled !== false) {
      if (decidedByComputer) {
        trace.push({
          rule_block: 'computer_use',
          outcome: result.decision,
          matched_rule: result.matched_rule,
          reason: result.reason,
          evaluated: true,
        });
      } else {
        trace.push({
          rule_block: 'computer_use',
          outcome: 'allow',
          matched_rule: undefined,
          reason: 'action allowed by computer_use rule',
          evaluated: true,
        });
      }
    } else {
      trace.push({
        rule_block: 'computer_use',
        outcome: 'skip',
        matched_rule: undefined,
        reason: 'rule disabled',
        evaluated: false,
      });
    }
  }
}

function appendSkippedRules(
  action: EvaluationAction,
  trace: RuleEvaluation[],
  reason: string,
): void {
  let blocks: string[];
  switch (action.type) {
    case 'tool_call':
      blocks = ['tool_access'];
      break;
    case 'egress':
      blocks = ['egress'];
      break;
    case 'file_read':
      blocks = ['forbidden_paths', 'path_allowlist'];
      break;
    case 'file_write':
      blocks = ['forbidden_paths', 'path_allowlist', 'secret_patterns'];
      break;
    case 'patch_apply':
      blocks = ['forbidden_paths', 'path_allowlist', 'patch_integrity'];
      break;
    case 'shell_command':
      blocks = ['shell_commands'];
      break;
    case 'computer_use':
      blocks = ['computer_use'];
      break;
    default:
      blocks = [];
      break;
  }

  for (const block of blocks) {
    trace.push({
      rule_block: block,
      outcome: 'skip',
      matched_rule: undefined,
      reason,
      evaluated: false,
    });
  }
}
